tests/lti_toolbox/test_launch_params.py

```diff
@@ -129,3 +126,152 @@ def _launch_params(lti_parameters):
+                "tool_consumer_instance_name": "Sandbox",
+                "tool_consumer_instance_description": "Moodle sandbox demo",
+                "accept_media_types": "application/vnd.ims.lti.v1.ltilink",
+                "accept_presentation_document_targets": "frame,iframe,window",
+                "accept_copy_advice": "false",
+                "accept_multiple": "true",
+                "accept_unsigned": "false",
+                "auto_create": "false",
+                "can_confirm": "false",
+                "content_item_return_url": "https://woop.com",
+                "title": (
+                    "Marsha LTI provider (never empty : fallback to moodle external tool name)"
+                ),
+                "text": "(current activity description if exists)",
+                "oauth_signature_method": "HMAC-SHA1",
+                "oauth_signature": "GEetrp41W4gCH5m1Fe6RPhf55W4=",
+            }
+        )
+
+    def test_urlencoded(self):
+        """Test urlencoded representation of an LTI launch request."""
+
+        selection_params = SelectionParams(
+            {
+                "lti_message_type": "ContentItemSelectionRequest",
+                "lti_version": "LTI-1p0",
+                "accept_media_types": "application/vnd.ims.lti.v1.ltilink",
+                "accept_presentation_document_targets": "frame,iframe,window",
+                "content_item_return_url": "http://testserver/",
+            }
+        )
+        expected = (
+            "lti_message_type=ContentItemSelectionRequest"
+            "&lti_version=LTI-1p0"
+            "&accept_media_types=application%2Fvnd.ims.lti.v1.ltilink"
+            "&accept_presentation_document_targets=frame%2Ciframe%2Cwindow"
+            "&content_item_return_url=http%3A%2F%2Ftestserver%2F"
+        )
+        self.assertEqual(expected, selection_params.urlencoded)
+
+    def test_invalid_parameter(self):
+        """Test behavior with invalid parameter in LTI launch request."""
+        with self.assertRaises(InvalidParamException):
+            SelectionParams(
+                {
+                    "lti_message_type": "ContentItemSelectionRequest",
+                    "lti_version": "LTI-1p0",
+                    "accept_media_types": "application/vnd.ims.lti.v1.ltilink",
+                    "accept_presentation_document_targets": "frame,iframe,window",
+                    "content_item_return_url": "http://testserver/",
+                    "invalid_param": "foo",
+                }
+            )
+
+    @staticmethod
+    def _selection_params(lti_parameters):
+        consumer = LTIConsumerFactory(slug="test_launch_params")
+        passport = LTIPassportFactory(title="test passport", consumer=consumer)
+        url = "http://testserver/lti/launch"
+        signed_parameters = sign_parameters(passport, lti_parameters, url)
+        return SelectionParams(signed_parameters)
```

Comment:
For code readability, I would recommend defining this on top of your class (at least before any test case).

Reply:
Totally agree, I kept it as the previous implementation. Lemme refactor both classes.